rasterio/_fill.pyx

```diff
@@ -26,38 +26,22 @@ def _fillnodata(image, mask, double max_search_distance=100.0,
 
     driver = GDALGetDriverByName("MEM")
 
-    if dtypes.is_ndarray(image):
-        # copy numpy ndarray into an in-memory dataset.
-        datasetname = str(uuid.uuid4()).encode('utf-8')
-        image_dataset = GDALCreate(
-            driver, <const char *>datasetname, image.shape[1], image.shape[0],
-            1, <GDALDataType>dtypes.dtype_rev[image.dtype.name], NULL)
-        image_band = GDALGetRasterBand(image_dataset, 1)
-        io_auto(image, image_band, True)
-    elif isinstance(image, tuple):
-        rdr = image.ds
-        band = (<DatasetReaderBase?>rdr).band(image.bidx)
-    else:
-        raise ValueError("Invalid source image")
-
-    if dtypes.is_ndarray(mask):
+    # copy numpy ndarray into an in-memory dataset.
+    datasetname = str(uuid.uuid4()).encode('utf-8')
+    image_dataset = GDALCreate(
```

Comment:
Any reason you are not using [`InMemoryRaster`](https://github.com/mapbox/rasterio/blob/master/rasterio/_io.pyx#L1572) here?

Reply:
Good question, I'll see if I can.

Reply:
Dang, that made a lot of code obsolete 🎉